Improve ILI9xxx display support for SPI to MCU interfaces

rexut opened the pull request:
In rare types of displays, discrete shift registers can be found to convert the SPI data stream to the ILI9xxx 16-bit MCU interface. Examples for that would be the ["Pico ResTouch LCD 3.5"](https://www.waveshare.com/wiki/Pico-ResTouch-LCD-3.5) shield or ["Pico Eval Board"](https://www.waveshare.com/wiki/Pico-Eval-Board) by Waveshare.

This PR extends the current driver stack with this new optional feature.

jfischer-no commented:
@rexut Thanks. There are efforts to move this driver to the "MIPI DBI driver class", see #64587, IIRC there are also plans to support 16 bit interfaces. Perhaps @danieldegrasse can comment on this.

danieldegrasse commented:
This is an interesting display- from what I can tell, it looks like the shift registers are being used to convert the SPI frames to 16 bit parallel RGB data? It looks like the final output format would be something like an 8080 interface, is that right?

I do think this kind of feature could be abstracted by the MIPI DBI API. Could you look at #64587 and see if you think a custom MIPI DBI driver would accomplish what you need here? It seems like the restriction this enforces is that all writes need to be 16 bits, whether the data sent is a command or display data

rexut commented:
> This is an interesting display- from what I can tell, it looks like the shift registers are being used to convert the SPI frames to 16 bit parallel RGB data? It looks like the final output format would be something like an 8080 interface, is that right?

Exactly. The ILI9xxx of the assembled LCD panel is hard-coded to talk via the 8080 interface. Waveshare has built a discrete shift register in front of it to access the LCD panel. This makes it an ordinary SPI device again for the outside world. BUT! any access to commands or data in the display RAM must be 16-bit wide.

> 
> I do think this kind of feature could be abstracted by the MIPI DBI API. Could you look at #64587 and see if you think a custom MIPI DBI driver would accomplish what you need here? It seems like the restriction this enforces is that all writes need to be 16 bits, whether the data sent is a command or display data

@danieldegrasse: I can try that. But I need to find some more free time for that. However, as I understand it, the MIPI DBI API is a "new" alternative for this type of displays in Zephyr and is still a work in progress. I will definitely have a look at that. Thanks for the hint.

@jfischer-no: Nevertheless, this PR now "only" extends the existing driver. This has already been main line, available and immediately usable for a long time. I could therefore imagine that this tweak will first be integrated into the existing systems via this PR and then the new MIPI DBI API will also be equipped with this feature.

danieldegrasse commented:
> Exactly. The ILI9xxx of the assembled LCD panel is hard-coded to talk via the 8080 interface. Waveshare has built a discrete shift register in front of it to access the LCD panel. This makes it an ordinary SPI device again for the outside world. BUT! any access to commands or data in the display RAM must be 16-bit wide.

I love the ingenuity

> @danieldegrasse: I can try that. But I need to find some more free time for that. However, as I understand it, the MIPI DBI API is a "new" alternative for this type of displays in Zephyr and is still a work in progress. I will definitely have a look at that. Thanks for the hint.

It is still a work in progress, but I *think* the implementation is relatively complete. I'm not opposed to adding support the way you are here, I just think that both approaches solve the same problem. The hope with the MIPI DBI API is that it can be more generic. I appreciate that this is a very one-off restriction though, and not really something that many displays would need. Are you aware of other displays with this kind of restriction?

rexut committed:
drivers: mipi_dbi: doc: fix clock in waveform diagram

In the Doxygen comments, the amount of SPI clocks now match the
number of data bits shown. The clock was shown half the speed.

Signed-off-by: Stephan Linz <linz@li-pro.net>

rexut commented:
> > Exactly. The ILI9xxx of the assembled LCD panel is hard-coded to talk via the 8080 interface. Waveshare has built a discrete shift register in front of it to access the LCD panel. This makes it an ordinary SPI device again for the outside world. BUT! any access to commands or data in the display RAM must be 16-bit wide.
> 
> I love the ingenuity
> 
> > @danieldegrasse: I can try that. But I need to find some more free time for that. However, as I understand it, the MIPI DBI API is a "new" alternative for this type of displays in Zephyr and is still a work in progress. I will definitely have a look at that. Thanks for the hint.
> 
> It is still a work in progress, but I _think_ the implementation is relatively complete. I'm not opposed to adding support the way you are here, I just think that both approaches solve the same problem. The hope with the MIPI DBI API is that it can be more generic. I appreciate that this is a very one-off restriction though, and not really something that many displays would need. Are you aware of other displays with this kind of restriction?

No, I don't know any further displays with such kind of special interconnection – I think, it's more a quirks to reuse cheap panels from stock than a intended system architecture.

However, the last few weeks I was busy with a Zephyr Meetup preparation and wasn't able to answer. Now I'm back to business and found some time to port this really special solution to the new MIPI DBI driver class for SPI.

What should I say – thank you for this new abstraction layer. From the original hack on ILI9xxx driver to this new MIPI DBI class extension I found a really better solution to integrate the correct byte stuffing – now without a preallocated 16-bit word data buffer. Before this was possible, because all ILI9xxx chips can only receive 32 byte at once. Now, with the more abstract MIPI DBI API, such kind of restriction have to avoid and the new Type C4 operating mode (SPI 4-wire 16-bit) as completely independent of amount of data to processed.

Please, have a look on the new code and tell me your opinion of improvements or better code. Thanks.

danieldegrasse commented:
Thanks for the work here. I'm slightly concerned about adding too many MIPI DBI modes to the header, since there are strictly only 3 modes defined by the spec. Could we instead add a devicetree prop to the MIPI DBI SPI driver that allows the user to specify a minimum write size? From my understanding, that is the only restriction this setup introduces versus a standard SPI interface.

rexut commented:
> Thanks for the work here. I'm slightly concerned about adding too many MIPI DBI modes to the header, since there are strictly only 3 modes defined by the spec. Could we instead add a devicetree prop to the MIPI DBI SPI driver that allows the user to specify a minimum write size? From my understanding, that is the only restriction this setup introduces versus a standard SPI interface.

Hm, good idea. I had something similar in mind at first, but then I didn't want to tinker too much with the MIPI DBI SPI driver. But if that's your way of thinking too, then let's do it that way. Give me a few more days (maybe weeks). I'll have to find some time between the daily tasks.

rexut commented:
Now I'm back from the "embedded world" in Nuremberg and can use the next few weeks for a revision. Can someone remove the "stall" label?

rexut commented:
The last change set has now been revised.

The former introduction of the new, non-standardized operating mode has been removed. Instead, there are now two new constant values for the minimum required transmission of bits via the SPI bus. This value can optionally be set to either 8 or 16 via the DTS as the "xfr-min-bits" property, with 8 being the default setting if nothing is defined in the DTS (default behaviour in mode MIPI_DBI_MODE_SPI_4WIRE). Like the operating modes, both constants are accessible via the header "dt-bindings/mipi_dbi/mipi_dbi.h" and can therefore be used via the new symbols MIPI_DBI_SPI_XFR_8BIT (default) or MIPI_DBI_SPI_XFR_16BIT in a DTS file.

Both new constants are interpreted in the functional part of the MIPI DBI SPI driver exclusively for the SPI 4 wire operating mode (type C3). The behavior of type C1 remains unchanged. The combination of MIPI_DBI_MODE_SPI_4WIRE and MIPI_DBI_SPI_XFR_8BIT (default setting) still corresponds exactly to the current behavior.

@danieldegrasse : let me know, if I'm completely wrong with the selected dt-binding file for the new "xfr-min-bits" property in "dts/bindings/mipi-dbi/zephyr,mipi-dbi-spi.yaml". For my own, this looks correct. The new property will describe the hardware interconnection similar to the correct wiring of reset and command/data selection GPIO and also if it would be possible to read back any data from the SPI device.

danieldegrasse commented on drivers/mipi_dbi/mipi_dbi_spi.c:
```diff
@@ -94,34 +98,103 @@ static int mipi_dbi_spi_write_helper(const struct device *dev,
+					data16 = sys_cpu_to_be16(data_buf[i]);
+
+					/* Set CD pin high for data */
+					gpio_pin_set_dt(&config->cmd_data, 1);
```
You likely only need to set the CD pin logic high here once, not in each loop iteration

rexut replied:
done

danieldegrasse commented on drivers/mipi_dbi/mipi_dbi_spi.c:
```diff
@@ -94,34 +98,103 @@ static int mipi_dbi_spi_write_helper(const struct device *dev,
+					buffer.len = sizeof(data16);
+
+					/* Set CD pin high for data */
+					gpio_pin_set_dt(&config->cmd_data, 1);
```
Same comment as above about moving the `gpio_pin_set_dt` call out of the for loop

rexut replied:
done

danieldegrasse commented on dts/bindings/mipi-dbi/zephyr,mipi-dbi-spi.yaml:
```diff
@@ -27,6 +27,22 @@ properties:
+      whose task is to convert the serial SPI bit stream to the parallel MCU
+      interface with clock and bit accuracy. Typically, these are 16 bits wide.
+
+      Use the macros not the actual enum value, here is the concordance list
```
```suggestion
      Use the macros, not the actual enum value. Here is the concordance list
```

Just a phrasing suggestion I received on the `mipi-mode` DT property, which has similar documentation wording

rexut replied:
Ah, I see. Done this and also fix the same issue in "mipi-dbi-spi-device.yaml" – see commit 120d144e326bb06f3adc589dd0d43de936ddaafe

danieldegrasse reviewed (commented):
Approach looks good to me- just a few minor comments

rexut committed:
dts: bindings: mipi-dbi-spi: fix language describing duplex

Update language describing mipi dbi spi duplex to be more clear.
Removing a run-on sentence.

Signed-off-by: Stephan Linz <linz@li-pro.net>

rexut commented on drivers/mipi_dbi/mipi_dbi_spi.c:
```diff
@@ -329,6 +402,8 @@ static struct mipi_dbi_driver_api mipi_dbi_spi_driver_api = {
 				    DT_INST_PHANDLE(n, spi_dev)),		\
 		    .cmd_data = GPIO_DT_SPEC_INST_GET_OR(n, dc_gpios, {}),	\
 		    .reset = GPIO_DT_SPEC_INST_GET_OR(n, reset_gpios, {}),	\
+		    .xfr_min_bits = DT_INST_PROP_OR(n, xfr_min_bits,            \
```
done

danieldegrasse commented on dts/bindings/mipi-dbi/mipi-dbi-spi-device.yaml:
```diff
@@ -14,7 +14,7 @@ properties:
       Selecting half duplex allows to use SPI MOSI as a bidirectional line,
       typically used when only one data line is connected.
-      Use the macros not the actual enum value, here is the concordance
+      Use the macros, not the actual enum value. Here is the concordance
```
Thanks for catching and fixing this as well :)

rexut commented:
@jfischer-no: as a collaborator of the display drivers (sadly no maintainer is listed) I would ask you for a +1 on review. As you mentioned, the display driver is now completely untouched. All the bitstream magic is done at MIPI DBI SPI level, one level above all display drivers. So theoretically it would be possible to support every display in the same way.

danieldegrasse commented:
@Dino-Li, I'm going to change myself to the assignee on this PR, as it modifies code in the MIPI DBI subsystem (which I maintain). If you feel this is incorrect, please go ahead and change the assignee back to yourself.

rexut commented:
@jfischer-no @Dino-Li @danieldegrasse : What are the next steps in the review/merge process? Is there anything else I need to do so that the PR will be accepted?

danieldegrasse commented:
@Dino-Li or @jfischer-no could you please provide a review on this PR so we can move this forwards?

danieldegrasse commented:
@ioannis-karachalios requested your review on this since you have to the MIPI DBI drivers, feel free to remove your request

jfischer-no commented on drivers/mipi_dbi/mipi_dbi_spi.c:
```diff
@@ -94,34 +98,107 @@ static int mipi_dbi_spi_write_helper(const struct device *dev,
+					data16 = sys_cpu_to_be16(data_buf[i]);
+
+					ret = spi_write(config->spi_dev, &dbi_config->config,
+							&buf_set);
+					if (ret < 0) {
+						goto out;
+					}
+				}
+			} else {
+				int stuffing = len % sizeof(data16);
+
+				/* Set CD pin high for data, if there are any */
+				if (len > 0) {
+					gpio_pin_set_dt(&config->cmd_data, 1);
+				}
+
+				/* pass through generic device data */
+				if (len - stuffing > 0) {
+					buffer.buf = (void *)data_buf;
+					buffer.len = len - stuffing;
+
+					ret = spi_write(config->spi_dev, &dbi_config->config,
+							&buf_set);
+					if (ret < 0) {
+						goto out;
+					}
+				}
+
+				/* iterate remaining data with stuffing */
+				for (int i = len - stuffing; i < len; i++) {
+					data16 = sys_cpu_to_be16(data_buf[i]);
+					buffer.buf = &data16;
+					buffer.len = sizeof(data16);
+
+					ret = spi_write(config->spi_dev, &dbi_config->config,
+							&buf_set);
+					if (ret < 0) {
+						goto out;
+					}
+				}
```
This block should be moved to separate spi_write_helper_4wire(), or even more finely subdivided.

rexut replied:
Makes sense, but then also for all other modes - currently 3wire and 4wire - otherwise you would give other developers no clear indication of the intention for breaking down when one mode is handled directly and another mode only with the help of an auxiliary function.

@danieldegrasse: What do you think, should I attach a separate commit with the two additional helper functions `spi_write_helper_3wire()` and `spi_write_helper_4wire()` to this PR, or just refactor the current commit and only provide the single helper function `spi_write_helper_4wire()` for the case of mode `MIPI_DBI_MODE_SPI_4WIRE`?

danieldegrasse replied:
If you are willing to split the code into a helper function `spi_write_helper_3wire` and `spi_write_helper4wire` in this PR, I would prefer that (a separate commit doing this refactoring prior to adding the 16 bit support would be best).

Given the approaching closure of the merge window though, I am ok if only the `spi_write_helper4wire` is done now, and the other change is made as a cleanup PR

rexut replied:
@danieldegrasse: With the new rebase, there are now two additional commits that split the SPI write function as well as the SPI read functions. In order to maintain consistency within the source, I think it makes sense to split the optional read as well. If this is too much for this PR, I can remove the one commit (read function) and submit it as a separate cleanup PR.

The forked helper functions now only contain the necessary SPI transfer algorithms. The decision as to which help function is necessary and that the device is locked and unlocked is still the responsibility of the SPI read/write function. The new help functions are specified as 'inline' (single point of use) and should potentially avoid a call stack overload.

However, two questions have now arisen with regard to the SPI read functions:

1. the previous implementation always called a `spi_release()` at the end, even if it is never necessary for `MIPI_DBI_MODE_SPI_3WIRE` (missing `SPI_LOCK_ON` or `SPI_HOLD_ON_CS` as with 4wire). This has been retained for the time being. But if it really was an unintentional error in the old code, then I will revise the one commit again.
2. in contrast to the `mipi_dbi_spi_write_helper()`, the `mipi_dbi_spi_command_read()` for the 3wire mode never checked whether an SPI 9-bit transfer is supported by the SPI controller at all. This check is now also included in the `mipi_dbi_spi_read_helper_3wire()`. Would that be okay?

rexut committed:
drivers: mipi_dbi_spi: splitting SPI read function

The more complex the SPI transfer algorithms become, the more
confusing the current implementation of the SPI command read
function becomes. Furthermore, if further as yet unknown MIPI
DBI modes are to be supported, the scope of this implementation
would increase dramatically. With the splitting now introduced,
the existing SPI transfer algorithms are moved to individual
auxiliary functions and the SPI read function only focus on
the decision of the respective MIPI DBI mode and the device
lock/unlock.

Signed-off-by: Stephan Linz <linz@li-pro.net>

rexut committed:
drivers: mipi_dbi_spi: splitting SPI write function

The more complex the SPI transfer algorithms become, the more
confusing the current implementation of the SPI write function
becomes. Furthermore, if further as yet unknown MIPI DBI modes
are to be supported, the scope of this implementation would
increase dramatically. With the splitting now introduced, the
existing SPI transfer algorithms are moved to individual
auxiliary functions and the SPI write function only focus on
the decision of the respective MIPI DBI mode and the device
lock/unlock.

Signed-off-by: Stephan Linz <linz@li-pro.net>

danieldegrasse commented on drivers/mipi_dbi/mipi_dbi_spi.c:
```diff
@@ -45,84 +49,228 @@ uint32_t var = MIPI_DBI_SPI_READ_REQUIRED;
+}
+
+static inline int
+mipi_dbi_spi_write_helper_4wire_16bit(const struct device *dev,
```
Could we add a check similar to `MIPI_DBI_SPI_READ_REQUIRED` to see if this function implementation is needed? Basically a check to see if any of the MIPI DBI SPI nodes in the devicetree use 16 bit transfers

rexut replied:
Done, also for the default 8 bit transfer. Because of situations where 16 bit would be the only required transfer mode, the 8 bit mode can be opt-out in same manner.

danieldegrasse commented:
Just a note- as the feature freeze is now active, this PR will have to go in after 3.7 release. Adding the correct milestone to reflect that

MaureenHelm commented:
@danieldegrasse please revisit

rexut commented:
@danieldegrasse I still have the outstanding improvements on the table, but won't get to them again until the fall. If the GH bot sets the issue to stall again, please mark it as active again - I'm on it.

rexut commented:
Hi @danieldegrasse. Thanks for reverting the "stale" flag for me once again. Yes, I'm still on here. Unfortunately, porting our downstream project Bridle to Zephyr v4 has been slowing me down massively since Vienna. I can only work on it in my limited spare time and a lot has happened in tooling, testing and documentation. But I think that the Christmas season will give me a little more flexibility. Thank you for your patience.

rexut committed:
drivers: mipi_dbi_spi: add 16-bit transfer to C4

Extends the MIPI DBI SPI driver class for operating mode C4, SPI 4-wire,
with 16 write clocks to send one or multiple byte for commands. Generic
data (e.g. GRAM) aligned to 16-bit are passed through and stuffed with
bytes if required.

Signed-off-by: Stephan Linz <linz@li-pro.net>

rexut committed:
tests: drivers: build_all: display: add config for 16-bit transfer

Add configuration with 16-bit transfer to C4 by the new `xfr-min-bits`
property. Such a setup generally only allows `write-only` access.

Signed-off-by: Stephan Linz <linz@li-pro.net>

danieldegrasse commented on tests/drivers/build_all/display/app.overlay:
```diff
@@ -161,6 +161,28 @@
+			#address-cells = <1>;
+			#size-cells = <0>;
+			xfr-min-bits = <MIPI_DBI_SPI_XFR_16BIT>;
+			write-only;
```
Minor nit- if we drop the `write-only` here, we would get better CI coverage, right?

rexut replied:
Well, basically the `write-only` property is irrelevant here for the resulting test binary, as the entire part of the read functions will be compiled in anyway by the `test_mipi_dbi_ili9342c` node above (without `write-only`). However, since in the context of the now new 16-bit support it is never possible to read, the explicit setting of `write-only` would not only be correct here at device tree level, but more necessary. This means that there is a notation right here in the test that should also be used in shields or boards for 16-bit accesses in order to avoid unnecessary code in the driver. The entire `MIPI_DBI_SPI_*_REQUIRED` macros already provide the correct opt-out strategy. The CI coverage should therefore not be disturbed or changed by this.

danieldegrasse replied:
Yeah, makes sense- I was thinking that the 16 bit support might disable reads based on the code here. Maybe it is worth documenting that 16 bit support and read support cannot coexist?

rexut replied:
Currently we (I) know only about the two rare types of displays `[referred in the beginning of this PR](https://github.com/zephyrproject-rtos/zephyr/pull/67089#issue-2061591661)`. Even this hardware never allows write only. That's why I add exactly that real existent case as additional test device. But that doesn't mean that everytime we see a display with 16 bit access also support read access. So I never document that explicit exclusion pn driver level. That exist only on real device level.

danieldegrasse commented on dts/bindings/mipi-dbi/zephyr,mipi-dbi-spi.yaml:
```diff
@@ -27,6 +27,22 @@ properties:
     description: |
       Reset GPIO pin. Set high to reset the display
 
+  xfr-min-bits:
```
One other (nonblocking) thought- we can use a string enum here, and use `DT_STRING_TOKEN` within the driver to access the property, similar to what was done here: #81293

danieldegrasse replied:
No worries, I am fine with the PR as is- we can always make this change in the future if desired, but it isn't a high priority to clean up (IMO)

danieldegrasse commented:
@jfischer-no could you take a look?

jfischer-no commented on drivers/mipi_dbi/mipi_dbi_spi.c:
```diff
@@ -38,6 +42,18 @@ struct mipi_dbi_spi_data {
 uint32_t var = MIPI_DBI_SPI_READ_REQUIRED;
 
+/* Expands to 1 if the node does reflect the enum in `xfr-min-bits` property */
+#define _XFR_8BITS(n) (DT_INST_PROP(n, xfr_min_bits) == MIPI_DBI_SPI_XFR_8BIT) |
+#define _XFR_16BITS(n) (DT_INST_PROP(n, xfr_min_bits) == MIPI_DBI_SPI_XFR_16BIT) |
```
IIRC underscores should not be used, see https://github.com/zephyrproject-rtos/zephyr/issues/9882, maybe just `XFR_.*BITS` or `MIPI_DBI_SPI_XFR_.*BITS`. _WRITE_ONLY_ABSENT should be fixed as well.
